src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharCategory.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Text.RegularExpressions
+{
+    internal class RegexCharCategory
+    {
+        private const byte Q = 5;    // quantifier
+        private const byte S = 4;    // ordinary stopper
+        private const byte Z = 3;    // ScanBlank stopper
+        private const byte X = 2;    // whitespace
+        private const byte E = 1;    // should be escaped
+
+        /// <summary>
+        /// For categorizing ASCII characters.
+        /// </summary>
+        private static readonly byte[] s_category = new byte[] {
```

Comment:
Instead of category char we could use an int index for second level table with target replacement chars.